src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/RewindDataIndexStore.cs

```diff
@@ -12,53 +12,43 @@ namespace Stratis.Bitcoin.Features.Consensus.ProvenBlockHeaders
     /// <inheritdoc />
     public class RewindDataIndexStore : IRewindDataIndexStore
     {
-        /// <summary>
-        /// The coin view to be used for getting rewind data.
-        /// </summary>
-        private readonly ICoinView coinView;
-
-        /// <summary>
-        /// The chained header tree.
-        /// </summary>
-        private readonly IChainedHeaderTree chainedHeaderTree;
-
         /// <summary>
         /// Internal cache for rewind data index. Key is a TxId + N (N is an index of output in a transaction)
         /// and value is a rewind data index.
         /// </summary>
         private readonly ConcurrentDictionary<string, int> items;
 
         /// <summary>
-        /// Number of items to keep in cache after the flush.
+        /// Number of blocks to keep in cache after the flush.
+        /// The number of items stored in cache is the sum of inputs used in every transaction in each of those blocks.
         /// </summary>
-        private const int NumberOfItemsToKeep = 500;
+        private int numberOfBlocksToKeep;
 
         /// <summary>
         /// Performance counter to measure performance of the save and get operations.
         /// </summary>
         private readonly BackendPerformanceCounter performanceCounter;
 
-        public RewindDataIndexStore(IDateTimeProvider dateTimeProvider, ICoinView coinView, IChainedHeaderTree chainedHeaderTree)
+        public RewindDataIndexStore(IDateTimeProvider dateTimeProvider)
         {
             Guard.NotNull(dateTimeProvider, nameof(dateTimeProvider));
-            Guard.NotNull(coinView, nameof(coinView));
-            Guard.NotNull(chainedHeaderTree, nameof(chainedHeaderTree));
 
             this.items = new ConcurrentDictionary<string, int>();
-            this.coinView = coinView;
-            this.chainedHeaderTree = chainedHeaderTree;
 
             this.performanceCounter = new BackendPerformanceCounter(dateTimeProvider);
         }
 
-        public async Task InitializeAsync()
+        /// <summary>
+        /// <inheritdoc />
+        public async Task InitializeAsync(IConsensus consensusParameters, ChainedHeader tip, ICoinView coinView)
         {
-            ChainedHeader tip = this.chainedHeaderTree.GetBestPeerTip();
-            int heightToSyncTo = tip.Height > NumberOfItemsToKeep ? tip.Height - NumberOfItemsToKeep : 0;
+            this.numberOfBlocksToKeep = (int)consensusParameters.MaxReorgLength;
```

Comment:
I am not entirely convinced we want to couple these together. If the rewind data size for the network starts getting very large (i.e. blocks are getting very full) you may not want to keep the entire cache in memory any longer, given that reorgs up to the maximum possible length are hopefully rare.

But perhaps there are good reasons to do it this way too.

Reply:
Well no, its essential to keep the number of max reorg because we may get branches that spend a reorged item at position CT- 499 and that may ultimately become the longest chain so we need to validate it as well.

Its rare yes but still possible